p2panda-stream/src/ordering/partial/operations.rs

```diff
@@ -0,0 +1,376 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::{collections::HashMap, marker::PhantomData};
+
+use p2panda_core::{Extensions, Hash, Operation};
+use p2panda_store::{LogStore, OperationStore};
+use thiserror::Error;
+
+use crate::ordering::{PartialOrder as InnerPartialOrder, PartialOrderError, PartialOrderStore};
+
+/// Struct for processing p2panda operations into a partial order based on dependencies expressed
+/// in their `previous` field.
+///
+/// This struct is a thin wrapper around ordering::PartialOrder struct which takes care of sorting
+/// the operation DAG into a partial order. Here we have the addition of a `LogStore` and
+/// `OperationStore` implementation (traits from `p2panda-store`) and an operation cache which
+/// helps us avoid unnecessary calls to the database.
+pub struct PartialOrder<L, E, OS, POS> {
+    /// A store containing p2panda operations.
+    operation_store: OS,
+
+    /// The inner PartialOrder struct which sorts the operation DAG into a partial order.
+    inner: InnerPartialOrder<Hash, POS>,
+
+    /// In memory cache of operations we have processed, before retrieving "ready" operations from
+    /// the store they will be taken from here if present. The store can be cleared at any time
+    /// in order to reduce memory use.
+    operation_cache: HashMap<Hash, Operation<E>>,
```

Comment:
A ring buffer would be cool here, with a configurable size on start up. I think `VecDeque` can be used as a ring buffer here.

Reply:
It's quite optimal that `HashMap` is fast for look-up, we'd lose this property with a `VecDeque` though. A capacity to the cache would be nice, it's of course possible to have your own logic to decide when to call `clear_cache()`.

Reply:
I'm sure there is ways to build a cache which is both performant to look up and has an automatic (configurable) limit